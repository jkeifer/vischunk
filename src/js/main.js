// Import styles
import '/src/styles.css';

// Import main application class
import { DataVisualizer } from '/src/js/visualization/visualizer.js';

// Initialize on load
document.addEventListener('DOMContentLoaded', () => {
    new DataVisualizer();
});
