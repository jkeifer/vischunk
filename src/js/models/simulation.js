import { LRUCache } from '/src/js/core/cache.js';
import { CONFIG } from '/src/js/core/constants.js';
import { GridCoordinate, CellCoordinate } from '/src/js/core/coordinates.js';

export class SimulationModel {
    constructor() {
        // Initialize caches with LRU policy
        this.linearizationCache = new LRUCache(CONFIG.LRU_DETAIL_CACHE_SIZE);
        this.positionCache = new LRUCache(CONFIG.LRU_DETAIL_CACHE_SIZE);
        this.chunkPositionCache = new LRUCache(CONFIG.LRU_CONFIG_CACHE_SIZE);
        this.cellPositionCache = new LRUCache(CONFIG.LRU_CONFIG_CACHE_SIZE);
        this.cellCoordinateCache = new Map();
    }

    // Create or get cached CellCoordinate system for given params
    getCellCoordinateSystem(params) {
        const [sizeX, sizeY, sizeZ = 1] = params.size;
        const [chunkX, chunkY, chunkZ = 1] = params.chunk;
        const { cellAlgorithm = 'row-major', chunkAlgorithm = 'row-major' } = params;

        const cacheKey = JSON.stringify([
            sizeX,
            sizeY,
            sizeZ,
            chunkX,
            chunkY,
            chunkZ,
            cellAlgorithm,
            chunkAlgorithm,
        ]);
        if (this.cellCoordinateCache.has(cacheKey)) {
            return this.cellCoordinateCache.get(cacheKey);
        }

        const chunksX = Math.ceil(sizeX / chunkX);
        const chunksY = Math.ceil(sizeY / chunkY);
        const chunksZ = Math.ceil(sizeZ / chunkZ);

        const chunkGrid = new GridCoordinate(
            chunksX,
            chunksY,
            chunksZ,
            chunkAlgorithm,
            this.chunkPositionCache
        );
        const cellCoordinate = new CellCoordinate(
            sizeX,
            sizeY,
            sizeZ,
            cellAlgorithm,
            chunkGrid,
            chunkX,
            chunkY,
            chunkZ,
            this.linearizationCache
        );

        this.cellCoordinateCache.set(cacheKey, cellCoordinate);
        return cellCoordinate;
    }

    linearizeCoordinate(x, y, z, sizeX, sizeY, sizeZ, algorithm) {
        const tempGrid = new GridCoordinate(
            sizeX,
            sizeY,
            sizeZ,
            algorithm,
            this.linearizationCache
        );
        return tempGrid.linearize(x, y, z);
    }

    getIntraChunkPosition(x, y, z, params) {
        const [sizeX, sizeY, sizeZ = 1] = params.size;
        const [chunkX, chunkY, chunkZ = 1] = params.chunk;
        const chunkCX = Math.floor(x / chunkX);
        const chunkCY = Math.floor(y / chunkY);
        const chunkCZ = Math.floor(z / chunkZ);
        const chunkStartX = chunkCX * chunkX;
        const chunkStartY = chunkCY * chunkY;
        const chunkStartZ = chunkCZ * chunkZ;
        const actualChunkX = Math.min(chunkStartX + chunkX, sizeX) - chunkStartX;
        const actualChunkY = Math.min(chunkStartY + chunkY, sizeY) - chunkStartY;
        const actualChunkZ = Math.min(chunkStartZ + chunkZ, sizeZ) - chunkStartZ;
        const localX = x - chunkStartX;
        const localY = y - chunkStartY;
        const localZ = z - chunkStartZ;
        return this.linearizeCoordinate(
            localX,
            localY,
            localZ,
            actualChunkX,
            actualChunkY,
            actualChunkZ,
            params.cellAlgorithm
        );
    }

    getInterChunkPosition(x, y, z, params) {
        const cellCoord = this.getCellCoordinateSystem(params);
        const chunk = cellCoord.getParentChunk(x, y, z);
        return cellCoord.chunkGrid.linearize(chunk.x, chunk.y, chunk.z);
    }

    getNormalizedChunkPosition(x, y, z, params) {
        return this.getInterChunkPosition(x, y, z, params);
    }

    getGlobalPosition(x, y, z, params) {
        const cellCoord = this.getCellCoordinateSystem(params);
        return cellCoord.getGlobalIndex(x, y, z);
    }

    getChunkBounds(chunkCX, chunkCY, chunkX, chunkY, sizeX, sizeY) {
        const startX = chunkCX * chunkX;
        const startY = chunkCY * chunkY;
        const endX = Math.min(startX + chunkX, sizeX);
        const endY = Math.min(startY + chunkY, sizeY);
        return { startX, startY, endX, endY };
    }

    getChunkGlobalRange(chunkCX, chunkCY, params) {
        const [sizeX, sizeY] = params.size;
        const [chunkX, chunkY] = params.chunk;
        const positions = [];
        for (let dy = 0; dy < chunkY; dy++) {
            for (let dx = 0; dx < chunkX; dx++) {
                const x = chunkCX * chunkX + dx;
                const y = chunkCY * chunkY + dy;
                if (x < sizeX && y < sizeY) {
                    positions.push(this.getGlobalPosition(x, y, 0, params));
                }
            }
        }
        if (positions.length === 0) {
            return null;
        }
        return {
            min: Math.min(...positions),
            max: Math.max(...positions),
            positions: positions,
        };
    }

    getChunkIndex(x, y, z, params) {
        const [chunkX, chunkY, chunkZ] = params.chunk;
        const cx = Math.floor(x / chunkX);
        const cy = Math.floor(y / chunkY);
        const cz = Math.floor(z / chunkZ);
        const [sizeX, sizeY] = params.size;
        const chunksX = Math.ceil(sizeX / chunkX);
        const chunksY = Math.ceil(sizeY / chunkY);
        return cx + cy * chunksX + cz * chunksX * chunksY;
    }

    calculateData(params) {
        const [sizeX, sizeY, sizeZ] = params.size;
        const totalCells = sizeX * sizeY * Math.max(1, sizeZ);
        const { requestedCells, touchedChunks } = this.calculateRequestedCellsAndChunks(
            params,
            sizeX,
            sizeY,
            sizeZ
        );
        const actualCells = this.calculateActualCellsFromChunks(
            touchedChunks,
            params,
            sizeX,
            sizeY,
            sizeZ
        );
        const chunkedRanges = this.calculateByteRanges(
            this.cellSetToPositions(actualCells, params)
        );
        const unchunkedRanges = this.calculateByteRanges(
            this.cellSetToPositions(requestedCells, params)
        );
        return {
            requestedCells,
            actualCells,
            touchedChunks,
            chunkedRanges,
            unchunkedRanges,
            totalCells,
        };
    }

    calculateRequestedCellsAndChunks(params, sizeX, sizeY, sizeZ) {
        const requestedCells = new Set();
        const touchedChunks = new Set();
        for (let x = params.query.x[0]; x <= Math.min(params.query.x[1], sizeX - 1); x++) {
            for (let y = params.query.y[0]; y <= Math.min(params.query.y[1], sizeY - 1); y++) {
                if (sizeZ > 0) {
                    for (
                        let z = params.query.z[0];
                        z <= Math.min(params.query.z[1], sizeZ - 1);
                        z++
                    ) {
                        requestedCells.add(`${x},${y},${z}`);
                        touchedChunks.add(this.getChunkIndex(x, y, z, params));
                    }
                } else {
                    requestedCells.add(`${x},${y},0`);
                    touchedChunks.add(this.getChunkIndex(x, y, 0, params));
                }
            }
        }
        return { requestedCells, touchedChunks };
    }

    calculateActualCellsFromChunks(touchedChunks, params, sizeX, sizeY, sizeZ) {
        const actualCells = new Set();
        const [chunkX, chunkY, _chunkZ] = params.chunk;
        const chunksX = Math.ceil(sizeX / chunkX);
        const chunksY = Math.ceil(sizeY / chunkY);
        touchedChunks.forEach(chunkIdx => {
            const chunkCoords = this.getChunkCoordsFromIndex(chunkIdx, chunksX, chunksY);
            this.addChunkCells(actualCells, chunkCoords, params, sizeX, sizeY, sizeZ);
        });
        return actualCells;
    }

    getChunkCoordsFromIndex(chunkIdx, chunksX, chunksY) {
        const cz = Math.floor(chunkIdx / (chunksX * chunksY));
        const cy = Math.floor((chunkIdx % (chunksX * chunksY)) / chunksX);
        const cx = chunkIdx % chunksX;
        return { cx, cy, cz };
    }

    addChunkCells(actualCells, chunkCoords, params, sizeX, sizeY, sizeZ) {
        const [chunkX, chunkY, chunkZ] = params.chunk;
        const { cx, cy, cz } = chunkCoords;
        for (let dx = 0; dx < chunkX; dx++) {
            for (let dy = 0; dy < chunkY; dy++) {
                for (let dz = 0; dz < (sizeZ > 0 ? chunkZ : 1); dz++) {
                    const x = cx * chunkX + dx;
                    const y = cy * chunkY + dy;
                    const z = cz * chunkZ + dz;
                    if (x < sizeX && y < sizeY && (sizeZ === 0 || z < sizeZ)) {
                        actualCells.add(`${x},${y},${z}`);
                    }
                }
            }
        }
    }

    cellSetToPositions(cellSet, params) {
        return [...cellSet].map(key => {
            const [x, y, z] = key.split(',').map(Number);
            return this.getGlobalPosition(x, y, z, params);
        });
    }

    calculateByteRanges(positions) {
        if (positions.length === 0) {
            return [];
        }
        const sorted = [...positions].sort((a, b) => a - b);
        const ranges = [];
        let start = sorted[0];
        let end = sorted[0];
        for (let i = 1; i < sorted.length; i++) {
            if (sorted[i] === end + 1) {
                end = sorted[i];
            } else {
                ranges.push([start, end]);
                start = end = sorted[i];
            }
        }
        ranges.push([start, end]);
        return ranges;
    }

    getOrCreateChunkColorMap(params, sizeX, sizeY, sizeZ) {
        const [chunkX, chunkY] = params.chunk;
        const chunksX = Math.ceil(sizeX / chunkX);
        const chunksY = Math.ceil(sizeY / chunkY);
        const chunksZ = Math.ceil(Math.max(1, sizeZ) / Math.max(1, params.chunk[2]));
        const cacheKey = JSON.stringify([chunksX, chunksY, chunksZ, params.chunkAlgorithm]);
        let chunkColorMap = this.positionCache.get(cacheKey);
        if (!chunkColorMap) {
            chunkColorMap = this.createChunkColorMap(params, chunksX, chunksY, chunksZ);
            this.positionCache.set(cacheKey, chunkColorMap);
        }
        return chunkColorMap;
    }

    createChunkColorMap(params, chunksX, chunksY, chunksZ) {
        const chunkPositions = [];
        for (let chunkCZ = 0; chunkCZ < chunksZ; chunkCZ++) {
            for (let chunkCY = 0; chunkCY < chunksY; chunkCY++) {
                for (let chunkCX = 0; chunkCX < chunksX; chunkCX++) {
                    const linearPos = this.linearizeCoordinate(
                        chunkCX,
                        chunkCY,
                        chunkCZ,
                        chunksX,
                        chunksY,
                        chunksZ,
                        params.chunkAlgorithm
                    );
                    chunkPositions.push({ chunkCX, chunkCY, chunkCZ, linearPos });
                }
            }
        }
        chunkPositions.sort((a, b) => a.linearPos - b.linearPos);
        const chunkColorMap = new Map();
        chunkPositions.forEach((chunk, index) => {
            const key = `${chunk.chunkCX},${chunk.chunkCY},${chunk.chunkCZ}`;
            chunkColorMap.set(key, index);
        });
        return chunkColorMap;
    }

    getOrCreatePositionToCellMap(params, sizeX, sizeY, sizeZ) {
        const [chunkSizeX, chunkSizeY] = params.chunk;
        const cacheKey = JSON.stringify([
            sizeX,
            sizeY,
            params.cellAlgorithm,
            params.chunkAlgorithm,
            chunkSizeX,
            chunkSizeY,
        ]);
        let positionToCell = this.positionCache.get(cacheKey);
        if (!positionToCell) {
            positionToCell = new Map();
            for (let z = 0; z < Math.max(1, sizeZ); z++) {
                for (let y = 0; y < sizeY; y++) {
                    for (let x = 0; x < sizeX; x++) {
                        const globalPos = this.getGlobalPosition(x, y, z, params);
                        positionToCell.set(globalPos, { x, y, z });
                    }
                }
            }
            this.positionCache.set(cacheKey, positionToCell);
        }
        return positionToCell;
    }

    getCellFromLinearIndex(cellIndex, params, sizeX, sizeY, chunkX, chunkY) {
        const [, , sizeZ] = params.size;
        const positionToCell = this.getOrCreatePositionToCellMap(params, sizeX, sizeY, sizeZ);
        const cell = positionToCell.get(cellIndex);

        if (!cell) {
            return null;
        }

        return {
            x: cell.x,
            y: cell.y,
            chunkX: Math.floor(cell.x / chunkX),
            chunkY: Math.floor(cell.y / chunkY),
        };
    }
}
