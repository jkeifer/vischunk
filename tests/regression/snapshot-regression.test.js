import { SimulationModel } from '../../src/js/models/simulation.js';
import { SNAPSHOT_CONFIGURATIONS } from '../fixtures/test-configurations.js';

describe('Snapshot-Based Regression Tests', () => {
    let simulation;
    
    beforeEach(() => {
        simulation = new SimulationModel();
    });

    describe('Algorithm Output Snapshots', () => {
        
        Object.entries(SNAPSHOT_CONFIGURATIONS).forEach(([key, config]) => {
            test(`${config.name} produces consistent output`, () => {
                const data = simulation.calculateData(config.params);
                
                // Create a deterministic snapshot object
                const snapshotData = {
                    // Basic metrics
                    metrics: {
                        totalCells: data.totalCells,
                        requestedCells: data.requestedCells.size,
                        actualCells: data.actualCells.size,
                        touchedChunks: data.touchedChunks.size,
                        rangeCount: data.chunkedRanges.length,
                        amplification: Number((data.actualCells.size / data.requestedCells.size).toFixed(4))
                    },
                    
                    // Test specific coordinate mappings for deterministic results
                    coordinateMappings: {},
                    
                    // Range structure (sorted for deterministic snapshots)
                    ranges: data.chunkedRanges.map(([start, end]) => ({ start, end })).sort((a, b) => a.start - b.start),
                    
                    // Sample of requested vs actual cells (for verification)
                    cellSamples: {
                        requestedSample: [...data.requestedCells].sort().slice(0, 10),
                        actualSample: [...data.actualCells].sort().slice(0, 10)
                    }
                };
                
                // Calculate global positions for test coordinates
                config.testCoordinates.forEach(([x, y, z]) => {
                    const globalPos = simulation.getGlobalPosition(x, y, z, config.params);
                    const intraChunkPos = simulation.getIntraChunkPosition(x, y, z, config.params);
                    const interChunkPos = simulation.getInterChunkPosition(x, y, z, config.params);
                    
                    snapshotData.coordinateMappings[`${x},${y},${z}`] = {
                        global: globalPos,
                        intraChunk: intraChunkPos,
                        interChunk: interChunkPos
                    };
                });
                
                // Sort coordinate mappings by key for deterministic snapshots
                const sortedMappings = {};
                Object.keys(snapshotData.coordinateMappings)
                    .sort()
                    .forEach(key => {
                        sortedMappings[key] = snapshotData.coordinateMappings[key];
                    });
                snapshotData.coordinateMappings = sortedMappings;
                
                expect(snapshotData).toMatchSnapshot();
            });
        });
    });

    describe('Cross-Algorithm Comparison Snapshots', () => {
        
        test('algorithm performance comparison matrix snapshot', () => {
            const baseConfig = {
                size: [16, 16, 1],
                chunk: [4, 4, 1],
                query: { x: [4, 8], y: [4, 8], z: [0, 0] }
            };
            
            const algorithms = ['row-major', 'col-major', 'z-order', 'hilbert'];
            const comparisonMatrix = {};
            
            algorithms.forEach(cellAlg => {
                algorithms.forEach(chunkAlg => {
                    const params = {
                        ...baseConfig,
                        cellAlgorithm: cellAlg,
                        chunkAlgorithm: chunkAlg
                    };
                    
                    const data = simulation.calculateData(params);
                    const key = `${cellAlg}â†’${chunkAlg}`;
                    
                    comparisonMatrix[key] = {
                        amplification: Number((data.actualCells.size / data.requestedCells.size).toFixed(4)),
                        coalescing: Number((data.touchedChunks.size / data.chunkedRanges.length).toFixed(4)),
                        ranges: data.chunkedRanges.length,
                        efficiency: Number((100 / (data.actualCells.size / data.requestedCells.size)).toFixed(2))
                    };
                });
            });
            
            // Sort the comparison matrix for deterministic snapshots
            const sortedMatrix = {};
            Object.keys(comparisonMatrix)
                .sort()
                .forEach(key => {
                    sortedMatrix[key] = comparisonMatrix[key];
                });
            
            expect(sortedMatrix).toMatchSnapshot();
        });
    });

    describe('Edge Case Snapshots', () => {
        
        test('boundary conditions produce consistent results', () => {
            const edgeCases = [
                {
                    name: 'single-cell-query',
                    params: {
                        size: [8, 8, 1],
                        chunk: [3, 3, 1],
                        cellAlgorithm: 'hilbert',
                        chunkAlgorithm: 'z-order',
                        query: { x: [4, 4], y: [4, 4], z: [0, 0] }
                    }
                },
                {
                    name: 'corner-query',
                    params: {
                        size: [10, 10, 1],
                        chunk: [4, 4, 1],
                        cellAlgorithm: 'z-order',
                        chunkAlgorithm: 'col-major',
                        query: { x: [0, 0], y: [0, 0], z: [0, 0] }
                    }
                },
                {
                    name: 'max-coordinate-query',
                    params: {
                        size: [12, 12, 1],
                        chunk: [5, 5, 1],
                        cellAlgorithm: 'row-major',
                        chunkAlgorithm: 'hilbert',
                        query: { x: [11, 11], y: [11, 11], z: [0, 0] }
                    }
                },
                {
                    name: 'partial-chunk-boundary',
                    params: {
                        size: [7, 9, 1],
                        chunk: [3, 4, 1],
                        cellAlgorithm: 'col-major',
                        chunkAlgorithm: 'row-major',
                        query: { x: [5, 6], y: [7, 8], z: [0, 0] }
                    }
                }
            ];
            
            const edgeCaseResults = {};
            
            edgeCases.forEach(({ name, params }) => {
                const data = simulation.calculateData(params);
                edgeCaseResults[name] = {
                    totalCells: data.totalCells,
                    requestedCells: data.requestedCells.size,
                    actualCells: data.actualCells.size,
                    touchedChunks: data.touchedChunks.size,
                    amplification: Number((data.actualCells.size / data.requestedCells.size).toFixed(4)),
                    ranges: data.chunkedRanges.map(([start, end]) => ({ start, end })).sort((a, b) => a.start - b.start)
                };
            });
            
            expect(edgeCaseResults).toMatchSnapshot();
        });
        
        test('3D configuration snapshots', () => {
            const config3D = {
                size: [4, 4, 4],
                chunk: [2, 2, 2],
                cellAlgorithm: 'hilbert',
                chunkAlgorithm: 'z-order',
                query: { x: [1, 2], y: [1, 2], z: [1, 2] }
            };
            
            const data = simulation.calculateData(config3D);
            
            // Test specific 3D coordinates
            const test3DCoords = [
                [0, 0, 0], [1, 1, 1], [2, 2, 2], [3, 3, 3],
                [0, 1, 2], [1, 2, 3], [2, 1, 0]
            ];
            
            const snapshot3D = {
                metrics: {
                    totalCells: data.totalCells,
                    requestedCells: data.requestedCells.size,
                    actualCells: data.actualCells.size,
                    touchedChunks: data.touchedChunks.size,
                    amplification: Number((data.actualCells.size / data.requestedCells.size).toFixed(4))
                },
                coordinateMappings: {}
            };
            
            test3DCoords.forEach(([x, y, z]) => {
                // Only test coordinates within bounds
                if (x < config3D.size[0] && y < config3D.size[1] && z < config3D.size[2]) {
                    const globalPos = simulation.getGlobalPosition(x, y, z, config3D);
                    const intraChunkPos = simulation.getIntraChunkPosition(x, y, z, config3D);
                    const interChunkPos = simulation.getInterChunkPosition(x, y, z, config3D);
                    
                    snapshot3D.coordinateMappings[`${x},${y},${z}`] = {
                        global: globalPos,
                        intraChunk: intraChunkPos,
                        interChunk: interChunkPos
                    };
                }
            });
            
            // Sort for deterministic snapshots
            const sortedMappings = {};
            Object.keys(snapshot3D.coordinateMappings)
                .sort()
                .forEach(key => {
                    sortedMappings[key] = snapshot3D.coordinateMappings[key];
                });
            snapshot3D.coordinateMappings = sortedMappings;
            
            expect(snapshot3D).toMatchSnapshot();
        });
    });
});