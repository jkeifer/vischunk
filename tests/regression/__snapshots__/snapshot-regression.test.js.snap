// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Snapshot-Based Regression Tests Algorithm Output Snapshots Basic Row-Major 16x16 produces consistent output 1`] = `
{
  "cellSamples": {
    "actualSample": [
      "10,10,0",
      "10,11,0",
      "10,4,0",
      "10,5,0",
      "10,6,0",
      "10,7,0",
      "10,8,0",
      "10,9,0",
      "11,10,0",
      "11,11,0",
    ],
    "requestedSample": [
      "4,4,0",
      "4,5,0",
      "4,6,0",
      "4,7,0",
      "4,8,0",
      "5,4,0",
      "5,5,0",
      "5,6,0",
      "5,7,0",
      "5,8,0",
    ],
  },
  "coordinateMappings": {
    "0,0,0": {
      "global": 0,
      "interChunk": 0,
      "intraChunk": 0,
    },
    "0,1,0": {
      "global": 4,
      "interChunk": 0,
      "intraChunk": 4,
    },
    "1,0,0": {
      "global": 1,
      "interChunk": 0,
      "intraChunk": 1,
    },
    "15,15,0": {
      "global": 255,
      "interChunk": 15,
      "intraChunk": 15,
    },
    "4,4,0": {
      "global": 80,
      "interChunk": 5,
      "intraChunk": 0,
    },
    "7,7,0": {
      "global": 95,
      "interChunk": 5,
      "intraChunk": 15,
    },
    "8,8,0": {
      "global": 160,
      "interChunk": 10,
      "intraChunk": 0,
    },
  },
  "metrics": {
    "actualCells": 64,
    "amplification": 2.56,
    "rangeCount": 2,
    "requestedCells": 25,
    "totalCells": 256,
    "touchedChunks": 4,
  },
  "ranges": [
    {
      "end": 111,
      "start": 80,
    },
    {
      "end": 175,
      "start": 144,
    },
  ],
}
`;

exports[`Snapshot-Based Regression Tests Algorithm Output Snapshots Column-Major 3D Configuration produces consistent output 1`] = `
{
  "cellSamples": {
    "actualSample": [
      "0,0,0",
      "0,0,1",
      "0,0,2",
      "0,0,3",
      "0,1,0",
      "0,1,1",
      "0,1,2",
      "0,1,3",
      "0,2,0",
      "0,2,1",
    ],
    "requestedSample": [
      "1,1,1",
      "1,1,2",
      "1,2,1",
      "1,2,2",
      "2,1,1",
      "2,1,2",
      "2,2,1",
      "2,2,2",
    ],
  },
  "coordinateMappings": {
    "0,0,0": {
      "global": 0,
      "interChunk": 0,
      "intraChunk": 0,
    },
    "0,0,1": {
      "global": 4,
      "interChunk": 0,
      "intraChunk": 4,
    },
    "0,1,0": {
      "global": 1,
      "interChunk": 0,
      "intraChunk": 1,
    },
    "1,0,0": {
      "global": 2,
      "interChunk": 0,
      "intraChunk": 2,
    },
    "1,1,1": {
      "global": 7,
      "interChunk": 0,
      "intraChunk": 7,
    },
    "2,2,2": {
      "global": 56,
      "interChunk": 7,
      "intraChunk": 0,
    },
    "3,3,3": {
      "global": 63,
      "interChunk": 7,
      "intraChunk": 7,
    },
  },
  "metrics": {
    "actualCells": 64,
    "amplification": 8,
    "rangeCount": 1,
    "requestedCells": 8,
    "totalCells": 64,
    "touchedChunks": 8,
  },
  "ranges": [
    {
      "end": 63,
      "start": 0,
    },
  ],
}
`;

exports[`Snapshot-Based Regression Tests Algorithm Output Snapshots Hilbert-ZOrder Mixed Algorithm produces consistent output 1`] = `
{
  "cellSamples": {
    "actualSample": [
      "0,0,0",
      "0,1,0",
      "0,2,0",
      "0,3,0",
      "0,4,0",
      "0,5,0",
      "0,6,0",
      "0,7,0",
      "1,0,0",
      "1,1,0",
    ],
    "requestedSample": [
      "2,2,0",
      "2,3,0",
      "2,4,0",
      "2,5,0",
      "3,2,0",
      "3,3,0",
      "3,4,0",
      "3,5,0",
      "4,2,0",
      "4,3,0",
    ],
  },
  "coordinateMappings": {
    "0,0,0": {
      "global": 0,
      "interChunk": 0,
      "intraChunk": 0,
    },
    "1,1,0": {
      "global": 2,
      "interChunk": 0,
      "intraChunk": 2,
    },
    "2,2,0": {
      "global": 8,
      "interChunk": 0,
      "intraChunk": 8,
    },
    "3,3,0": {
      "global": 10,
      "interChunk": 0,
      "intraChunk": 10,
    },
    "4,4,0": {
      "global": 48,
      "interChunk": 3,
      "intraChunk": 0,
    },
    "7,7,0": {
      "global": 58,
      "interChunk": 3,
      "intraChunk": 10,
    },
  },
  "metrics": {
    "actualCells": 64,
    "amplification": 4,
    "rangeCount": 1,
    "requestedCells": 16,
    "totalCells": 64,
    "touchedChunks": 4,
  },
  "ranges": [
    {
      "end": 63,
      "start": 0,
    },
  ],
}
`;

exports[`Snapshot-Based Regression Tests Algorithm Output Snapshots Large Array Representative Sample produces consistent output 1`] = `
{
  "cellSamples": {
    "actualSample": [
      "10,10,0",
      "10,11,0",
      "10,12,0",
      "10,13,0",
      "10,14,0",
      "10,15,0",
      "10,16,0",
      "10,17,0",
      "10,18,0",
      "10,19,0",
    ],
    "requestedSample": [
      "10,10,0",
      "10,11,0",
      "10,12,0",
      "10,13,0",
      "10,14,0",
      "10,15,0",
      "10,16,0",
      "10,8,0",
      "10,9,0",
      "11,10,0",
    ],
  },
  "coordinateMappings": {
    "0,0,0": {
      "global": 0,
      "interChunk": 0,
      "intraChunk": 0,
    },
    "15,15,0": {
      "global": 191,
      "interChunk": 2,
      "intraChunk": 63,
    },
    "16,16,0": {
      "global": 512,
      "interChunk": 8,
      "intraChunk": 0,
    },
    "24,24,0": {
      "global": 640,
      "interChunk": 10,
      "intraChunk": 0,
    },
    "31,31,0": {
      "global": 703,
      "interChunk": 10,
      "intraChunk": 63,
    },
    "7,7,0": {
      "global": 63,
      "interChunk": 0,
      "intraChunk": 63,
    },
    "8,8,0": {
      "global": 128,
      "interChunk": 2,
      "intraChunk": 0,
    },
  },
  "metrics": {
    "actualCells": 256,
    "amplification": 3.1605,
    "rangeCount": 3,
    "requestedCells": 81,
    "totalCells": 1024,
    "touchedChunks": 4,
  },
  "ranges": [
    {
      "end": 191,
      "start": 128,
    },
    {
      "end": 575,
      "start": 448,
    },
    {
      "end": 895,
      "start": 832,
    },
  ],
}
`;

exports[`Snapshot-Based Regression Tests Cross-Algorithm Comparison Snapshots algorithm performance comparison matrix snapshot 1`] = `
{
  "col-major→col-major": {
    "amplification": 2.56,
    "coalescing": 2,
    "efficiency": 39.06,
    "ranges": 2,
  },
  "col-major→hilbert": {
    "amplification": 2.56,
    "coalescing": 1.3333,
    "efficiency": 39.06,
    "ranges": 3,
  },
  "col-major→row-major": {
    "amplification": 2.56,
    "coalescing": 2,
    "efficiency": 39.06,
    "ranges": 2,
  },
  "col-major→z-order": {
    "amplification": 2.56,
    "coalescing": 1,
    "efficiency": 39.06,
    "ranges": 4,
  },
  "hilbert→col-major": {
    "amplification": 2.56,
    "coalescing": 2,
    "efficiency": 39.06,
    "ranges": 2,
  },
  "hilbert→hilbert": {
    "amplification": 2.56,
    "coalescing": 1.3333,
    "efficiency": 39.06,
    "ranges": 3,
  },
  "hilbert→row-major": {
    "amplification": 2.56,
    "coalescing": 2,
    "efficiency": 39.06,
    "ranges": 2,
  },
  "hilbert→z-order": {
    "amplification": 2.56,
    "coalescing": 1,
    "efficiency": 39.06,
    "ranges": 4,
  },
  "row-major→col-major": {
    "amplification": 2.56,
    "coalescing": 2,
    "efficiency": 39.06,
    "ranges": 2,
  },
  "row-major→hilbert": {
    "amplification": 2.56,
    "coalescing": 1.3333,
    "efficiency": 39.06,
    "ranges": 3,
  },
  "row-major→row-major": {
    "amplification": 2.56,
    "coalescing": 2,
    "efficiency": 39.06,
    "ranges": 2,
  },
  "row-major→z-order": {
    "amplification": 2.56,
    "coalescing": 1,
    "efficiency": 39.06,
    "ranges": 4,
  },
  "z-order→col-major": {
    "amplification": 2.56,
    "coalescing": 2,
    "efficiency": 39.06,
    "ranges": 2,
  },
  "z-order→hilbert": {
    "amplification": 2.56,
    "coalescing": 1.3333,
    "efficiency": 39.06,
    "ranges": 3,
  },
  "z-order→row-major": {
    "amplification": 2.56,
    "coalescing": 2,
    "efficiency": 39.06,
    "ranges": 2,
  },
  "z-order→z-order": {
    "amplification": 2.56,
    "coalescing": 1,
    "efficiency": 39.06,
    "ranges": 4,
  },
}
`;

exports[`Snapshot-Based Regression Tests Edge Case Snapshots 3D configuration snapshots 1`] = `
{
  "coordinateMappings": {
    "0,0,0": {
      "global": 0,
      "interChunk": 0,
      "intraChunk": 0,
    },
    "0,1,2": {
      "global": 33,
      "interChunk": 4,
      "intraChunk": 1,
    },
    "1,1,1": {
      "global": 6,
      "interChunk": 0,
      "intraChunk": 6,
    },
    "1,2,3": {
      "global": 55,
      "interChunk": 6,
      "intraChunk": 7,
    },
    "2,1,0": {
      "global": 9,
      "interChunk": 1,
      "intraChunk": 1,
    },
    "2,2,2": {
      "global": 56,
      "interChunk": 7,
      "intraChunk": 0,
    },
    "3,3,3": {
      "global": 62,
      "interChunk": 7,
      "intraChunk": 6,
    },
  },
  "metrics": {
    "actualCells": 64,
    "amplification": 8,
    "requestedCells": 8,
    "totalCells": 64,
    "touchedChunks": 8,
  },
}
`;

exports[`Snapshot-Based Regression Tests Edge Case Snapshots boundary conditions produce consistent results 1`] = `
{
  "corner-query": {
    "actualCells": 16,
    "amplification": 16,
    "ranges": [
      {
        "end": 15,
        "start": 0,
      },
    ],
    "requestedCells": 1,
    "totalCells": 100,
    "touchedChunks": 1,
  },
  "max-coordinate-query": {
    "actualCells": 4,
    "amplification": 4,
    "ranges": [
      {
        "end": 123,
        "start": 120,
      },
    ],
    "requestedCells": 1,
    "totalCells": 144,
    "touchedChunks": 1,
  },
  "partial-chunk-boundary": {
    "actualCells": 20,
    "amplification": 5,
    "ranges": [
      {
        "end": 55,
        "start": 40,
      },
      {
        "end": 62,
        "start": 59,
      },
    ],
    "requestedCells": 4,
    "totalCells": 63,
    "touchedChunks": 4,
  },
  "single-cell-query": {
    "actualCells": 9,
    "amplification": 9,
    "ranges": [
      {
        "end": 35,
        "start": 27,
      },
    ],
    "requestedCells": 1,
    "totalCells": 64,
    "touchedChunks": 1,
  },
}
`;
